### PR TITLE
RTED -- first PR

### PR DESCRIPTION
**Pull request by snim2**
This is the first of several PRs implementing the RTED algorithm of Pawlik and Augsten (2011).

This PR includes some API changes, the RTED config struct and its constructor.

### CODE REVIEW

**Comment by snim2** [comment]
Now needs squashing :(

File: src/lib/info_tree.rs [hunk 1]
```diff
@@ -52,11 +53,22 @@ const REL_SUBTREES_SIZE: usize = 3;
 // Size of the node type vector.
 const NODE_TYPE_SIZE: usize = 3;
 
-enum PathIdx {
+/// Constants related to forward and reverse paths through trees.
+pub enum PathIdx {
+    #[allow(missing_docs)]
```
Comment (ltratt) [inline src/lib/info_tree.rs]: If you push this `#[allow]` on the `enum` does it suppress the warning for each entry or ...?

File: src/lib/rted.rs [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+    // Distances between every pair of subtrees in the form
+    // delta(F, G) - delta(F', G') which is at most 1.0.
+    delta_bit: Vec<Vec<u8>>,
+    // Stores a forest preorder for a given `i` and `j`.
+    ij: Vec<Vec<usize>>,
+    cost_v: Vec<Vec<Vec<usize>>>,
+    cost_w: Vec<Vec<usize>>,
+    // T array from Demaine's algorithm, stores delta(Fv,Gij), v on heavy path.
+    t: Vec<Vec<f64>>,
+    // Copy values from the `t` vector., in case in single compute period values
```
Comment (ltratt) [inline src/lib/rted.rs]: Stray `.`?

File: src/lib/rted.rs [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+        for _ in 0..max_size {
+            rted.ij.push(vec![0; max_size]);
+        }
+        for _ in 0..rted.size_src {
+            rted.delta.push(vec![0.0; rted.size_dst]);
+        }
+        for _ in 0..rted.size_src {
+            rted.delta_bit.push(vec![0; rted.size_dst]);
+        }
+        for i in 0..3 {
```
Comment (ltratt) [inline src/lib/rted.rs]: Magic constant should be defined somewhere.

File: src/lib/rted.rs [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+            rted.cost_v.push(vec![]);
+            for _ in 0..rted.size_src {
```
Comment (ltratt) [inline src/lib/rted.rs]: Ditto.

File: src/lib/rted.rs [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+                rted.cost_v[i].push(vec![0; rted.size_dst]);
+            }
+        }
+        for _ in 0..3 {
```
Comment (ltratt) [inline src/lib/rted.rs]: Ditto.

File: src/lib/rted.rs [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+        let labels_src = &self.itree_src.get_info_vec(InfoIdx::Post2Label);
+        let labels_dst = &self.itree_dst.get_info_vec(InfoIdx::Post2Label);
+        let sizes_src = &self.itree_src.get_info_vec(InfoIdx::Post2Size);
+        let sizes_dst = &self.itree_dst.get_info_vec(InfoIdx::Post2Size);
+        for x in 0..sizes_src.len() {
+            for y in 0..sizes_dst.len() {
+                if labels_src[x] == labels_dst[y] {
+                    self.delta_bit[x][y] = 0;
+                } else {
+                    // If this set, the labels differ, cost of relabelling is
```
Comment (ltratt) [inline src/lib/rted.rs]: Comment is a bit hard to parse.

**Comment by ltratt** [comment]
My comments are superficial: I'd like to reduce the number of doc suppression things though. They really are quite ugly...

File: src/lib/info_tree.rs (hunk 1 above)
Comment (snim2) [reply src/lib/info_tree.rs]: Yes, it doesn't affect the rest of the file.

**Comment by snim2** [comment]
OK, ready for re-review!

File: src/lib/rted.rs [hunk 7]
```diff
@@ -68,7 +68,7 @@ pub struct RTEDConfig<'a> {
     cost_w: Vec<Vec<usize>>,
     // T array from Demaine's algorithm, stores delta(Fv,Gij), v on heavy path.
     t: Vec<Vec<f64>>,
-    // Copy values from the `t` vector., in case in single compute period values
+    // Copy values from the `t` vector, in case in single compute period values
```
Comment (ltratt) [inline src/lib/rted.rs]: I still can't parse "...in case in single...".

File: src/lib/rted.rs [hunk 8]
```diff
@@ -154,8 +154,8 @@ impl<'a> RTEDConfig<'a> {
                 if labels_src[x] == labels_dst[y] {
                     self.delta_bit[x][y] = 0;
                 } else {
-                    // If this set, the labels differ, cost of relabelling is
-                    // set to cost_match.
+                    // If this set, labels in the two trees differ and the cost
```
Comment (ltratt) [inline src/lib/rted.rs]: I can't parse this either. What's "If this set, labels in ..." refer to?

File: src/lib/rted.rs [hunk 9]
```diff
@@ -71,8 +71,9 @@ pub struct RTEDConfig<'a> {
     cost_w: Vec<Vec<usize>>,
     // T array from Demaine's algorithm, stores delta(Fv,Gij), v on heavy path.
     t: Vec<Vec<f64>>,
-    // Copy values from the `t` vector, in case in single compute period values
-    // are  overwritten before they are read due to a change of forest ordering.
+    // Copy values from the `t` vector. This is used when, in a single compute
+    // period, values are overwritten before they are read, due to a change of
+    // forest ordering.
```
Comment (snim2) [inline src/lib/rted.rs]: Note that the "compute period" refers to a function that hasn't been committed yet.

**Comment by snim2** [comment]
OK, ready again.

**Comment by ltratt** [comment]
Please squash.

**Comment by snim2** [comment]
Squashed.
